From ee5f1d8af0eec06f185b8511b9b11192f802148b Mon Sep 17 00:00:00 2001
From: Debian LibreOffice Maintainers <debian-openoffice@lists.debian.org>
Date: Fri, 17 Oct 2014 23:53:24 +0200
Subject: [PATCH] optimize_performance_after_replaceall_searchall.diff

  * Optimize performance after replaceall & searchall  (LP: #1342175)
    - upstream 1cf19ea84794ca065749667b480dfed2d27d47b7
    - upstream 1e721077b43de84edab2a3ed2f316ddcbec6e3ec
    - upstream 91502a72c12c559442e8bf77c27a516b49c2a68d

===================================================================
---
 sc/inc/column.hxx                     |  3 ++-
 sc/inc/columnspanset.hxx              |  3 +++
 sc/source/core/data/column2.cxx       |  4 ++--
 sc/source/core/data/columnspanset.cxx | 14 +++++++++++---
 sc/source/core/data/table2.cxx        | 30 ++++++++++++++++++++++--------
 sc/source/core/data/table3.cxx        |  4 +++-
 6 files changed, 43 insertions(+), 15 deletions(-)

diff --git a/sc/inc/column.hxx b/sc/inc/column.hxx
index 66cd524..9911772 100644
--- a/sc/inc/column.hxx
+++ b/sc/inc/column.hxx
@@ -24,6 +24,7 @@
 #include "global.hxx"
 #include "address.hxx"
 #include "rangenam.hxx"
+#include "rangelst.hxx"
 #include "types.hxx"
 #include "mtvelements.hxx"
 #include "formula/types.hxx"
@@ -255,7 +256,7 @@ public:
     ScAttrIterator* CreateAttrIterator( SCROW nStartRow, SCROW nEndRow ) const;
 
     void UpdateSelectionFunction(
-        const ScMarkData& rMark, ScFunctionData& rData, ScFlatBoolRowSegments& rHiddenRows );
+        const ScRangeList& rRanges, ScFunctionData& rData, ScFlatBoolRowSegments& rHiddenRows );
 
     void CopyToColumn(
         sc::CopyToDocContext& rCxt, SCROW nRow1, SCROW nRow2, sal_uInt16 nFlags, bool bMarked,
diff --git a/sc/inc/columnspanset.hxx b/sc/inc/columnspanset.hxx
index c12f13e..af1cc6e 100644
--- a/sc/inc/columnspanset.hxx
+++ b/sc/inc/columnspanset.hxx
@@ -20,6 +20,7 @@ class ScDocument;
 class ScColumn;
 class ScMarkData;
 class ScRange;
+class ScRangeList;
 
 namespace sc {
 
@@ -124,6 +125,8 @@ public:
      */
     void scan(const ScMarkData& rMark, SCTAB nTab, SCCOL nCol);
 
+    void scan(const ScRangeList& rRanges, SCTAB nTab, SCCOL nCol);
+
     void set(SCROW nRow1, SCROW nRow2, bool bVal);
 
     void getRows(std::vector<SCROW> &rRows) const;
diff --git a/sc/source/core/data/column2.cxx b/sc/source/core/data/column2.cxx
index a8ba063..d64e575 100644
--- a/sc/source/core/data/column2.cxx
+++ b/sc/source/core/data/column2.cxx
@@ -3352,10 +3352,10 @@ public:
 
 //  multiple selections:
 void ScColumn::UpdateSelectionFunction(
-    const ScMarkData& rMark, ScFunctionData& rData, ScFlatBoolRowSegments& rHiddenRows )
+    const ScRangeList& rRanges, ScFunctionData& rData, ScFlatBoolRowSegments& rHiddenRows )
 {
     sc::SingleColumnSpanSet aSpanSet;
-    aSpanSet.scan(rMark, nTab, nCol); // mark all selected rows.
+    aSpanSet.scan(rRanges, nTab, nCol); // mark all selected rows.
 
     // Exclude all hidden rows.
     ScFlatBoolRowSegments::RangeData aRange;
diff --git a/sc/source/core/data/columnspanset.cxx b/sc/source/core/data/columnspanset.cxx
index 7bc517b..b46033b 100644
--- a/sc/source/core/data/columnspanset.cxx
+++ b/sc/source/core/data/columnspanset.cxx
@@ -275,9 +275,17 @@ void SingleColumnSpanSet::scan(const ScMarkData& rMark, SCTAB nTab, SCCOL nCol)
         return;
 
     ScRangeList aRanges = rMark.GetMarkedRanges();
-    for (size_t i = 0, n = aRanges.size(); i < n; ++i)
-    {
-        const ScRange* p = aRanges[i];
+    scan(aRanges, nTab, nCol);
+}
+
+void SingleColumnSpanSet::scan(const ScRangeList& rRanges, SCTAB nTab, SCCOL nCol)
+{
+    for (size_t i = 0, n = rRanges.size(); i < n; ++i)
+     {
+        const ScRange* p = rRanges[i];
+        if (nTab < p->aStart.Tab() || p->aEnd.Tab() < nTab)
+            continue;
+
         if (nCol < p->aStart.Col() || p->aEnd.Col() < nCol)
             // This column is not in this range. Skip it.
             continue;
diff --git a/sc/source/core/data/table2.cxx b/sc/source/core/data/table2.cxx
index 1a37801..e700987 100644
--- a/sc/source/core/data/table2.cxx
+++ b/sc/source/core/data/table2.cxx
@@ -1897,10 +1897,17 @@ bool ScTable::HasAttrib( SCCOL nCol1, SCROW nRow1, SCCOL nCol2, SCROW nRow2, sal
 
 bool ScTable::HasAttribSelection( const ScMarkData& rMark, sal_uInt16 nMask ) const
 {
-    bool bFound = false;
-    for (SCCOL i=0; i<=MAXCOL && !bFound; i++)
-        bFound |= aCol[i].HasAttribSelection( rMark, nMask );
-    return bFound;
+    std::vector<sc::ColRowSpan> aSpans = rMark.GetMarkedColSpans();
+
+    for (size_t i = 0; i < aSpans.size(); ++i)
+    {
+        for (SCCOLROW j = aSpans[i].mnStart; j < aSpans[i].mnEnd; ++j)
+        {
+            if (aCol[j].HasAttribSelection(rMark, nMask))
+                return true;
+        }
+    }
+    return false;
 }
 
 
@@ -2171,10 +2178,17 @@ bool ScTable::HasBlockMatrixFragment( SCCOL nCol1, SCROW nRow1, SCCOL nCol2, SCR
 
 bool ScTable::HasSelectionMatrixFragment( const ScMarkData& rMark ) const
 {
-    bool bFound = false;
-    for (SCCOL i=0; i<=MAXCOL && !bFound; i++)
-        bFound |= aCol[i].HasSelectionMatrixFragment(rMark);
-    return bFound;
+    std::vector<sc::ColRowSpan> aSpans = rMark.GetMarkedColSpans();
+
+    for ( size_t i=0; i<aSpans.size(); i++ )
+    {
+        for ( SCCOLROW j=aSpans[i].mnStart; j<aSpans[i].mnEnd; j++ )
+        {
+            if ( aCol[j].HasSelectionMatrixFragment(rMark) )
+                return true;
+        }
+    }
+    return false;
 }
 
 
diff --git a/sc/source/core/data/table3.cxx b/sc/source/core/data/table3.cxx
index 946416d..4711356 100644
--- a/sc/source/core/data/table3.cxx
+++ b/sc/source/core/data/table3.cxx
@@ -61,6 +61,7 @@
 #include <listenercontext.hxx>
 #include <sharedformula.hxx>
 #include <refhint.hxx>
+#include "rangelst.hxx"
 #include <listenerquery.hxx>
 #include <bcaslot.hxx>
 
@@ -2999,12 +3000,13 @@ xub_StrLen ScTable::GetMaxNumberStringLen(
 
 void ScTable::UpdateSelectionFunction( ScFunctionData& rData, const ScMarkData& rMark )
 {
+    ScRangeList aRanges = rMark.getMarkedRanges();
     for (SCCOL nCol = 0; nCol <= MAXCOL && !rData.bError; ++nCol)
     {
         if (pColFlags && ColHidden(nCol))
             continue;
 
-        aCol[nCol].UpdateSelectionFunction(rMark, rData, *mpHiddenRows);
+        aCol[nCol].UpdateSelectionFunction(aRanges, rData, *mpHiddenRows);
     }
 }
 
-- 
1.9.1

